--
title: "ADSP WGS GWAS results of 10 AD-related genes"
author: "Xulong Wang"
date: "January 13, 2017"
output: html_document
---

+ Variants: MAF > 0.01
+ GWAS model: generalized linear model (not GLMM)

10 AD-related genes

```{r}

genes = c("BCHE", "TREM2", "PDGFA", "CLU", "CR1", "ABCA7", "IL1RAP", "TYROBP", "STAT3", "STAT4")

load("~/Dropbox/GitHub/X/hg19.rdt")
hg19 = hg19[hg19$NAME %in% genes, ]
hg19

```

Variants of the 10 genes 

```{r}

load("/data/xwang/adsp3/meta.rdt")

meta = meta[meta$CHR %in% unique(hg19$CHR), ]
meta = meta[meta$MAF > 0.01, ]
meta$POS = as.numeric(meta$POS)
meta$ID = as.character(meta$ID)

meta = lapply(genes, function(gene) { cat(gene, "\n")
  chr = hg19[match(gene, hg19$NAME), "CHR"]
  start = hg19[match(gene, hg19$NAME), "START"]
  end = hg19[match(gene, hg19$NAME), "END"]
  
  meta[meta$CHR == chr & meta$POS > start - 1e5 & meta$POS < end + 1e5, ]
}); names(meta) = genes

sapply(meta, nrow)

meta = do.call(rbind, meta)
meta$Gene = gsub("\\..*", "", rownames(meta))
rownames(meta) = NULL

```

GWAS results of the variants

```{r}

load("/data/xwang/adsp3/glm.rdt")
glm = glm[rownames(glm) %in% meta$ID, ]

results = cbind(meta, glm[match(meta$ID, rownames(glm)), ])
rownames(results) = NULL

results = results[c("ID", "CHR", "POS", "REF", "ALT", "MAF", "Gene", "p", "se", "P")]
save(results, file = "~/Dropbox/GitHub/wgs2/genes/10genes.rdt")

library(xlsx)
file = "/home/xwang/Dropbox/GitHub/wgs2/genes/AD10.xlsx"
write.xlsx(results, file = file, sheetName = "GLM")

```
